corehq/apps/reports/standard/sms.py

```diff
@@ -322,49 +319,123 @@ def uses_locations(self):
+            # We can chain ANDs together, but not ORs, so we have to do all the ORs at the same time.
+            if incl_survey and incl_other:
+                filters = (workflow_filter | survey_filter | other_filter)
+            elif incl_survey:
+                filters = (workflow_filter | survey_filter)
+            elif incl_other:
+                filters = (workflow_filter | other_filter)
+            else:
+                filters = workflow_filter
+            return data_.filter(filters)
+
+        def order_by_col(data_):
+            col_fields = ['date', 'couch_recipient', 'phone_number', 'direction', 'text']
+            sort_col = self.request_params.get('iSortCol_0')
+            if sort_col is not None and sort_col < len(col_fields):
+                data_ = data_.order_by(col_fields[sort_col])
+            if self.request_params.get('sSortDir_0') == 'desc':
+                data_ = data_.reverse()
+            return data_
+
+        def get_phone_number(phone_number):
+            if abbreviate_phone_number and phone_number is not None:
+                return phone_number[0:7] if phone_number[0] == "+" else phone_number[0:6]
+            return phone_number
+
+        get_direction = lambda d: {INCOMING: _('Incoming'), OUTGOING: _('Outgoing')}.get(d, '---')
+
+        def get_timestamp(date_):
+            timestamp = ServerTime(date_).user_time(self.timezone).done()
+            return timestamp.strftime(SERVER_DATETIME_FORMAT)
+
+        def get_contact_link(couch_recipient, couch_recipient_doc_type):
+            doc_info = self.get_recipient_info(couch_recipient_doc_type, couch_recipient, contact_cache)
+            if doc_info:
+                username = doc_info.display
+                url = doc_info.link
+                if doc_info.is_deleted:
+                    username = '%s (%s %s)' % (username, _('Deleted'), _(doc_info.type_display))
+                    url = None
+            else:
+                username = None
+                url = None
+            username = username or "-"
+            if url:
+                return '<a target="_blank" href="{url}">{username}</a>'.format(username=username, url=url)
+            return username
 
-            doc_info = self.get_recipient_info(message.couch_recipient_doc_type,
-                message.couch_recipient, contact_cache)
+        startdate = json_format_datetime(self.datespan.startdate_utc)
+        enddate = json_format_datetime(self.datespan.enddate_utc)
+        data = SMS.objects.filter(
+            domain=self.domain,
+            date__range=(startdate, enddate),
+        ).exclude(
+            # Exclude outgoing messages that have not yet been processed
+            direction=OUTGOING,
+            processed=False
+        )
+        data = filter_by_types(data)
+        data = order_by_col(data)
+        data = data[self.pagination.start:self.pagination.start + self.pagination.count]
```

Comment:
is there a way to do the pagination without querying for the whole result set first?